ext/magic/extconf.rb

```diff
@@ -46,6 +156,90 @@ def windows?
   RbConfig::CONFIG['target_os'] =~ /mswin|mingw32|windows/
 end
 
+def truffle?
+  ::RUBY_ENGINE == 'truffleruby'
+end
+
+def concat_flags(*args)
+  args.compact.join(" ")
+end
+
+def do_help
+  print(MAGIC_HELP_MESSAGE)
+  exit!(0)
+end
+
+def do_clean
+  root = Pathname(PACKAGE_ROOT_DIR)
+  pwd  = Pathname(Dir.pwd)
+
+  # Skip if this is a development work tree
+  unless (root + '.git').exist?
+    message("Cleaning files only used during build.\n")
+
+    # (root + 'tmp') cannot be removed at this stage because
+    # libmagic.so is yet to be copied to lib.
+
+    # clean the ports build directory
+    Pathname.glob(pwd.join('tmp', '*', 'ports')) do |dir|
+      FileUtils.rm_rf(dir, verbose: true)
+    end
+
+    FileUtils.rm_rf(root + 'ports' + 'archives', verbose: true)
+
+    # Remove everything but share/ directory
+    remove_paths = %w[bin include]
+    remove_paths << 'lib' if config_static?
+
+    Pathname.glob(File.join(root, 'ports', '*', 'libmagic', '*')) do |dir|
+      remove_paths.each do |path|
+        remove_dir = File.join(dir, path)
+        FileUtils.rm_rf(remove_dir, verbose: true)
+      end
+    end
+  end
+
+  exit!(0)
+end
+
+#
+#  main
+#
+do_help if arg_config('--help')
+do_clean if arg_config('--clean')
+
+if config_system_libraries?
+  message "Building ruby-magic using system libraries.\n"
+
+  dir_config('magic')
+else
+  message "Building ruby-magic using packaged libraries.\n"
+
+  static_p = config_static?
+  message "Static linking is #{static_p ? 'enabled' : 'disabled'}.\n"
+  cross_build_p = config_cross_build?
+  message "Cross build is #{cross_build_p ? 'enabled' : 'disabled'}.\n"
+
+  libmagic_recipe = process_recipe('libmagic', LIBMAGIC_TAG, static_p, cross_build_p) do |recipe|
+    recipe.files = [{
+                      url: "https://ruby-magic.s3.eu-central-1.amazonaws.com/file-#{recipe.version}.tar.gz",
+                      sha256: LIBIMAGE_SHA256
+                    }]
+  end
+
+  $LIBPATH = [File.join(libmagic_recipe.path, 'lib')]
+  $CFLAGS << " -I#{File.join(libmagic_recipe.path, 'include')} "
+
+  if static_p
+    ENV['PKG_CONFIG_PATH'] = "#{libmagic_recipe.path}/lib/pkgconfig"
+    # mkmf appends -- to the first option
+    $LIBS += " " + pkg_config('libmagic', 'libs --static')
+    $LDFLAGS.gsub!('-lmagic', '')
+    $LIBS.gsub!('-lmagic', '')
+    $LIBS += " " + File.join(libmagic_recipe.path, 'lib', "libmagic.#{$LIBEXT}")
+  end
```

Comment:
When building our shared object to by linked dynamically, we could also set `rpath` to ensure that other version of `libmagic` will never be picked. For instance, either:

```ruby
$LDFLAGS += " -Wl,-rpath -Wl,#{recipe.path}/lib"
```

or perhaps:

```ruby
$LDFLAGS += " -Wl,-rpath,#{File.expand_path(recipe.work_path)}"
```

Not sure which one would be better to use. What do you think?

Reply:
I think the first option works since the work path may be cleaned up later when `--disable-static` is used.

Reply:
We *probably* could omit:

```
$LIBS += " " + File.join(libmagic_recipe.path, 'lib', "libmagic.#{$LIBEXT}")
```

But I think this assures that a static library is linked in case the shared library is only built.

Reply:
Sounds good!